### ISSUE TITLE
full typing support for validate_call as well as easy access to raw calls

### ISSUE THREAD

**Issue by RonnyPfannschmidt** [pydantic/pydantic#9883]
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

### Description

while working with openapi-codegen generated apis, i foudn that validate_call decorated apis are quite a pain when wanting to be trypesafe and/or invoking them

i want to propose using a fully typed callable protocol to typecheck them and including a way to correctly invoke the raw function even for bound methods

this would require to

1. turn [ValidateCallWrapper](https://github.com/pydantic/pydantic/blob/d654a0766c2f3c6fe0a12718f32aa3bf4d3ecc86/pydantic/_internal/_validate_call.py#L15) into a generic callable
2. introducing a `ValidateMethodWrapper`providing bound method behaviour for it
3. providing both of those wrappers a function to do a unchecked call that shares the underlying signatures but works correctly

that way something like the following would be possible:

```python
from pydantic import validate_call

class MyApi:
   "utterly trimmed down example"
   @validate_call
   def get_username(self, userid: int) -> str:
      return str(userid)



def test_what_i_want():
   api = MyApi()
   reveal_type(api.get_username.unchecked("bad_data_test")
  

```

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [X] [Data validation/parsing](https://docs.pydantic.dev/concepts/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/concepts/serialization/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/concepts/json_schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/concepts/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/concepts/config/)
- [ ] [Field Types](https://docs.pydantic.dev/api/types/) - adding or changing a particular data type
- [X] [Function validation decorator](https://docs.pydantic.dev/concepts/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/concepts/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/concepts/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by Viicos** [comment]
> while working with openapi-codegen generated apis, i foudn that validate_call decorated apis are quite a pain when wanting to be trypesafe and/or invoking them

Could you give an example where it isn't typesafe?

`validate_call` is currently typed as (omitting the first overload):

https://github.com/pydantic/pydantic/blob/d654a0766c2f3c6fe0a12718f32aa3bf4d3ecc86/pydantic/validate_call_decorator.py#L24-L25

So from a type checking perspective, it returns the same decorated callable untouched.

Regarding the `.unchecked` call, you can already do it this way:

```python
from pydantic import validate_call

@validate_call
def test(a: int) -> None:
    pass

test.raw_function(a="a")  # OK at runtime
```

However, type checkers can't know about `raw_function`. As you suggested, this would require exposing `ValidateCallWrapper` (or any other new class) in some way, and make it generic over a `ParamSpec` and a `TypeVar` for the return type:

```python
P = ParamSpec("P")
T = TypeVar("T")

class ValidateCallWrapper(Generic[P, T]):
    ...

    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> T: ...

    def raw_function(self, *args: P.args, **kwargs: P.kwargs) -> T: ...


def validate_call(func: Callable[P, T]) -> ValidateCallWrapper[P, T]: ...

@validate_call
def test(a: int) -> None:
    pass

test(a=1)  # type checks
test.raw_function(a=1)  # type checks as well
```

However, you will hit issues when decorating methods, as type checkers aren't going to understand when the decorated method is bound or not.

The same issue can be found for [`functools.cache`](https://docs.python.org/3/library/functools.html#functools.cache) in typeshed, and there's currently no ideal solution for this, see:
- https://github.com/python/typeshed/pull/11662
- https://github.com/python/typeshed/pulls?q=functools+paramspec, showing a bunch of issues/PRs related to this.

---

So I think your best solution right now is to

**Comment by RonnyPfannschmidt** [comment]
@Viicos in my example i was using a decorated method, where the raw function does not automatically get bound

additional the raw function is not typed and missing from the type signature

**Comment by Viicos** [comment]
> additional the raw function is not typed and missing from the type signature

As I described, there's no way to currently type check `raw_function`. We'll have to wait for the semantics of `ParamSpec` to be updated so that it can support both methods and function (or unbound methods). Additional discussion can be found [here](https://discuss.python.org/t/allow-self-binding-for-generic-paramspec/50948), and doesn't seem trivial.

**Comment by RonnyPfannschmidt** [comment]
i worked it out in https://github.com/python/mypy/issues/16200#issuecomment-2220227424 already as i needed a similar helper with extra properties and technically speaking its incorrect to pretend a callable protocol object is/can be the same as a plain function (there are numerous callable objects that do not bind, and sometimes intentionally/rightfully so)

**Comment by Viicos** [comment]
Well you can still experiment with it and see if it works, considering the following should type check:

```python
@validate_call
def test(a: int = 1) -> None: ...

class A:
    @validate_call
    def test(self, a: int = 1) -> None: ...

test(a=1) 
test()
test.raw_function(a=1)
test.raw_function()

A().test(a=1)
A().test()
A().test.raw_function(a=1)
A().test.raw_function()

A.test(A(), a=1)
A.test(A())
A.test.raw_function(A(), a=1)
A.test.raw_function(A())
```

But I'm pretty certain you will face the same issues as `functools.cache`. In particular, I experimented with here https://github.com/python/typeshed/pull/11767

**Comment by Viicos** [comment]
Closing for now as we can't fully express the functionality with the current typing constructs. We decided to type `validate_call` as:

https://github.com/pydantic/pydantic/blob/5e241c6514757083b4376d5924ae8bfc52a02c44/pydantic/validate_call_decorator.py#L25

So that we preserve type checking of arguments, at the expense of not having the `.raw_function` exposed.

`functools.cache` is doing it the other way around, exposing methods (such as `cache_clear`) on the wrapper but dropping the arguments type checking. It's an unfortunate tradeoff but considering the negligible usage of `.raw_function` I believe we are doing the right choice here.
